for-upstream-only/nixos/profiles/cachix-use-export.nix

```diff
@@ -0,0 +1,119 @@
+      default = null;
+      description = ''
+        A JSON Lines file produced by the cachix export command. It
+        will be read during evaluation. This can be a path expression, which
+        will not be loaded into the Nix store by the declaring module.
+      '';
+    };
+    deployedSecretsPath = lib.mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = ''
+        The path to the deployed nix.cachix.secretsFile on the target
+        machine(s). This should be a plain string
+        literal, to avoid accidentally copying secrets into the Nix store.
+
+        The file will be read by root and any pull credentials will be made
+        available exclusively to the Nix daemon. Non-root users of Nix tools
+        may need to provide the credentials themselves to function properly.
+        For this reason the caches are merely trusted and not enabled by default.
+      '';
+    };
+    /*
+      TODO: make this actually configurable
+            This module is currently limited to the hardened version which
+            deploys netrc to daemon-netrc with 0400, root:root
+
+    rootOnly = lib.mkOption {
+      type = types.bool;
+      default = true;
+      description = ''
+        Writes the pull tokens (if any) to a file that is only readable by root
+        and the Nix daemon. This makes it harder for users to retrieve these
+        tokens.
+      '';
+    };
+    */
+
+  };
+
+  config = lib.mkIf ( # || because of the assertions
+                     cfg.secretsFile != null || cfg.deployedSecretsPath != null
+                    ) {
+
+    assertions = [
+      { assertion = cfg.deployedSecretsPath != null -> cfg.secretsFile != null;
+        message = ''
+          You need to specify nix.cachix.secretsFile in order to provide the
+          non-sensitive parts of the cache configuration.
+
+          WARNING: If you've used a path expression in cfg.deployedSecretsPath,
+          you may want to delete it from your Nix store!
+        '';
+      }
+
+      # TODO: not required when file is not sensitive.
+      { assertion = cfg.secretsFile != null -> cfg.deployedSecretsPath != null;
+        message = ''
+          You need to deploy the Cachix secrets file to the machine outside the
+          Nix store and set nix.cachix.deployedSecretsPath to the location of the
+          deployed file.
+        '';
+      }
+    ];
+
+    nix.trustedBinaryCaches = ["https://cache.nixos.org"] ++ map (o: "https://${o.cacheName}.cachix.org") pubkeys;
+    nix.binaryCachePublicKeys = map (o: o.publicKey) pubkeys;
+
+    nix.extraOptions = ''
+      netrc-file = /etc/nix/daemon-netrc
+    '';
+
+    systemd.paths.cachix-secrets = {
+      wantedBy = [ "multi-user.target" ];
+      pathConfig.PathExists = cfg.deployedSecretsPath;
+      pathConfig.PathChanged = cfg.deployedSecretsPath;
+      pathConfig.Unit = "cachix-install-netrc.service";
+    };
+
+    systemd.services.cachix-install-netrc = {
```

Comment:
Doesn't this duplicate `withCaches` logic? We have to decide if we want to install these globally or just for the agent.

I'd prefer that only the agent uses the configuration, as there may be multiple agents or the target platform in the future won't have NixOS modules so it will be much harder to setup an agent.

Reply:
The nix daemon also needs access to the keys. Narinfo is fetched by the nix client, whereas the nar is fetched by the daemon.

Deduplicating it would require world-readable secrets, which this solution doesn't.
If some platforms call for it, we can still choose a simplified configuration with world-readable secrets for those, yet keep most other users on safer agents.

Reply:
Follow up at https://github.com/hercules-ci/hercules-ci-agent/issues/72